File: src/TicTacToe/haskell/Data/TicTacToe/GameResult.hs
-- | A game result is one of
--
-- * Player 1 wins
--
-- * Player 2 wins
--
-- * Neither player wins (draw)
module Data.TicTacToe.GameResult
(
  GameResult
-- * Reduction (fold)
, gameResult
, playerGameResult
-- * Construction
, win
, player1Wins
, player2Wins
, draw
-- * Decisions
, isPlayer1Wins
, isPlayer2Wins
, isDraw
) where

import Data.TicTacToe.Player

-- | A game result.
data GameResult =
  Win Player
  | Draw
  deriving (Eq, Show)

-- | Fold a game result.
gameResult ::
  (Player -> x) -- ^ If either of the players won.
  -> x -- ^ If the game was a draw.
  -> GameResult -- ^ The game result to fold.
  -> x
gameResult win _    (Win p) =
  win p
gameResult _   draw Draw    =
  draw

-- | Fold a game result.
playerGameResult ::
  x -- ^ If player 1 won.
  -> x -- ^ If player 2 won.
  -> x -- ^ If the game was a draw.
  -> GameResult -- ^ The game result to fold.
  -> x
playerGameResult p1 p2 =
  gameResult (player p1 p2)

-- | Construct a game result with a win for the given player.
win ::
  Player -- ^ The player to win.
  -> GameResult
win =
  Win

-- | Construct a game result with a win for player 1.
player1Wins ::
  GameResult
player1Wins =
  Win player1

-- | Construct a game result with a win for player 2.
player2Wins ::
  GameResult
player2Wins =
  Win player2

-- | Construct a game result that is a draw.
draw ::
  GameResult
draw =
  Draw

-- | Returns whether or not player 1 won for the game result.
isPlayer1Wins ::
  GameResult
  -> Bool
isPlayer1Wins =
  playerGameResult True False False

-- | Returns whether or not player 2 won for the game result.
isPlayer2Wins ::
  GameResult
  -> Bool
isPlayer2Wins =
  playerGameResult False True False

-- | Returns whether the game result is a draw.
isDraw ::
  GameResult
  -> Bool
isDraw =
  playerGameResult False False True

File: tictactoe/java/src/tictactoe/GameResult.java
package tictactoe;

import fj.F;
import fj.data.Option;

import static fj.data.Option.some;
import static tictactoe.Player.Player1;
import static tictactoe.Player.Player2;

public enum GameResult {
  Player1Wins, Player2Wins, Draw;

  public boolean isWin() {
    return this == Player1Wins || this == Player2Wins;
  }

  public boolean isDraw() {
    return !isWin();
  }

  public Option<Player> winner() {
    return this == Player1Wins ?
             some(Player1) :
             this == Player2Wins ?
               some(Player2) :
               Option.<Player>none();
  }

  public <X> X strictFold(final X player1Wins, final X player2Wins, final X draw) {
    return this == Player1Wins ?
             player1Wins :
             this == Player2Wins ?
               player2Wins :
               draw;
  }

  @Override
  public String toString() {
    return winner().option("draw", new F<Player, String>() {
      @Override
      public String f(final Player p) {
        return p.toString() + " wins";
      }
    });
  }

  public static GameResult win(final Player p) {
    return p == Player1 ? Player1Wins : Player2Wins;
  }
}
